interfaces/builtin/netlink_protocol.go

```diff
@@ -0,0 +1,92 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/snap"
+)
+
+const netlinkProtocolSummary = `allows communication through the kernel custom netlink protocol`
+
+const netlinkProtocolBaseDeclarationSlots = `
+  netlink-protocol:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
+    deny-auto-connection: true
+`
+
+// netlinkProtocolInterface type
+type netlinkProtocolInterface struct {
+	commonInterface
+}
+
+// BeforePrepareSlot checks the slot definition is valid
+func (iface *netlinkProtocolInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// Must have a protocol number
+	number, ok := slot.Attrs["protocol"]
+	if !ok {
+		return fmt.Errorf("netlink-protocol slot must have a protocol number attribute")
+	}
+
+	// Valid values of number
+	if _, ok := number.(int64); !ok {
+		return fmt.Errorf("netlink-protocol slot protocol number attribute must be an int")
```

Comment:
is there some kind of known range for such custom protocols, if there is this should check that protocol is in that range. I don't think we want to make this a way to access kernel predefined protocols.

Reply:
there are 24 (0-23) existing protocols: http://162.213.32.110:8080/source/xref/ubuntu-core20/libnl3/include/linux-private/linux/netlink.h?r=a6071805#7
I'm not aware of upper limit now apart of int limit